src/jdk.compiler/share/classes/com/sun/tools/javac/parser/JavacParser.java

```diff
@@ -933,16 +933,29 @@ JCExpression term2Rest(JCExpression t, int minprec) {
                 int pos = token.pos;
                 nextToken();
                 int typePos = token.pos;
-                JCExpression type = parseType();
+                JCModifiers mods = optFinal(0);
+                JCExpression type = unannotatedType(false);
                 JCTree pattern;
                 if (token.kind == IDENTIFIER) {
                     checkSourceLevel(token.pos, Feature.PATTERN_MATCHING_IN_INSTANCEOF);
-                    JCModifiers mods = F.at(Position.NOPOS).Modifiers(0);
                     JCVariableDecl var = toP(F.at(token.pos).VarDef(mods, ident(), type, null));
                     TreeInfo.getStartPos(var);
                     pattern = toP(F.at(typePos).BindingPattern(var));
                     TreeInfo.getStartPos(pattern);
                 } else {
+                    if (mods.annotations.nonEmpty()) {
```

Comment:
We don't seem to check for the source level here? E.g. is it possible that `instanceof @Foo Bar` will not issue the usual error if compiled with source != 16?

Reply:
Oops. Having `instanceof @Foo Bar` is valid, due to type annotations. See e.g.:
https://github.com/openjdk/jdk/blob/d2b66196b4509a32613f476d7247f207289a75e3/test/langtools/tools/javac/annotations/typeAnnotations/TargetTypes.java#L102

But, when I was adding support for the final modifier on binding variables, I've forgot to disable the final modifier on non-pattern matching `instanceof`, so code like `o instanceof final String` was accepted, which is, of course, wrong.

This should be fixed by:
https://github.com/openjdk/jdk/pull/1192/commits/a737f9229adf86af90fdab98a67ad009d76ab44c